frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1521 @@
+					// should call `migrate_child`).
+					// 3. this is the root of a child key, and we are finishing all child-keys (and
+					// should call `migrate_top`).
+
+					// NOTE: this block is written intentionally to verbosely for easy of
+					// verification.
+					match (
+						top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX),
+						self.prev_tick_child,
+					) {
+						(false, false) => {
+							// continue the top key migration
+							self.migrate_top();
+						},
+						(true, false) => {
+							self.last_child = Some(Default::default());
+							self.migrate_child();
+							self.prev_tick_child = true;
+						},
+						(true, true) => {
+							// we're done with migrating a child-root.
+							self.prev_tick_child = false;
+							self.migrate_top();
+						},
+						(false, true) => {
+							// should never happen.
+							log!(error, "LOGIC ERROR: unreachable code [0].");
+							Pallet::<T>::halt();
+						},
+					};
+				},
+				(None, Some(_)) => {
+					log!(error, "LOGIC ERROR: unreachable code [1].");
+					Pallet::<T>::halt()
+				},
+				(None, None) => {
+					// nada
+				},
+			}
+		}
+
+		/// Migrate the current child key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_child(&mut self) {
+			use sp_io::default_child_storage as child_io;
+			if self.last_child.is_none() || self.last_top.is_none() {
+				// defensive: this function is only called when both of these values exist.
+				// much that we can do otherwise..
+				frame_support::traits::defensive_path("cannot migrate child key.");
+				return
+			}
+
+			let last_child = self.last_child.as_ref().expect("value checked to be `Some`; qed");
+			let last_top = self.last_top.clone().expect("value checked to be `Some`; qed");
+
+			let child_root = Pallet::<T>::transform_child_key_or_halt(&last_top);
+			let maybe_current_child = child_io::next_key(child_root, last_child);
+			if let Some(ref current_child) = maybe_current_child {
+				let added_size = if let Some(data) = child_io::get(child_root, &current_child) {
+					child_io::set(child_root, current_child, &data);
+					data.len() as u32
+				} else {
+					Zero::zero()
+				};
+				self.dyn_size = self.dyn_size.saturating_add(added_size);
+				self.dyn_child_items.saturating_inc();
+			}
+
+			log!(trace, "migrated a child key, next_child_key: {:?}", maybe_current_child);
+			self.last_child = maybe_current_child;
+		}
+
+		/// Migrate the current top key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_top(&mut self) {
+			if self.last_top.is_none() {
+				// defensive: this function is only called when this value exist.
+				// much that we can do otherwise..
+				frame_support::traits::defensive_path("cannot migrate top key.");
+				return
+			}
+			let last_top = self.last_top.as_ref().expect("value checked to be `Some`; qed");
```

Comment:
```suggestion
			let last_top = match self.last_top { 
				Some(l) => l,
				None => {
					// defensive: this function is only called when this value exist.
					// much that we can do otherwise..
					frame_support::traits::defensive_path("cannot migrate top key.");
					return
				}
			};
```